[AIP-217] Unreachable resources (#259)
googleapis/aip: aip/0217.md

---
aip:
  id: 217
  state: reviewing
  created: 2019-08-26
permalink: /217
redirect_from:
  - /0217
---

# Unreachable resources

Occasionally, a user may ask for a list of resources, and some set of resources
in the list are temporarily unavailable. For example, a user may ask to list
resources across multiple locations, but one of those locations is temporarily
down. In this situation, it is still desirable to give the user as much as
possible, while indicating that something is missing.

## Guidance

If a method to retrieve data is capable of partially failing due to one or more
resources being temporarily unreachable, the response message **must** include
a field to indicate this:

```proto
// The request message for ListBooks.
message ListBooksRequest {
  // The publisher to list books from.
  // To list books across all publishers, use "publishers/-".
  string parent = 1;

>   // Other fields...

Reviewer: // Other field as defined in AIP-132

Author: ```suggestion
// Other fields (as defined in AIP-132)...
```

}

message ListBooksResponse {
  // The books matching the request.
  repeated Book books = 1;

  // The next page token, if there are more books matching the
  // request.
  string next_page_token = 2;

  // Unreachable resources. If any books that would match the
  // request were not included (for example, because one publisher's

>   // API was down), the unreachable resources are listed here.

Reviewer: This is a bit ambiguous. Does this list include the "unavailable books" or the publishers? We need to be explicit about what resources will end up in here.

Author: I ended up just removing most of this comment, and rely on the bullets below. I could not find a concise but clear way to word it.

  repeated string unreachable = 3;
}
```

- The field **must** be a repeated string, and **should** be named
  `unreachable`.
- The field **must** be set to the names of the resources which are the cause

>   of the issue, such as the locations or resources that could not be reached.

Reviewer: Ambiguous -- see note above. We should explicitly note that in some cases, parents are unavailable, in whcih case, the parents should be listed here as "unreachable". In other cases, we know that a resource exists but are unable to retrieve the data, in which case, those unreachable resources should be listed.

Reviewer 2: Additionally, I think we should be clear about how homogeneous or otherwise these might be.

Suppose we have `clusters/x/nodes/y/tasks/z` and we're listing tasks:

- Cluster A has nodes A1 and A2 which are fine, so we can list tasks in them
- Cluster B has nodes B1 and B2, but B2 is failing. We only include tasks in B1
- Cluster C is down entirely

Would the unreachable list be "clusters/b/nodes/b1" and "clusters/c", or should it just be "clusters/b" and "clusters/c"?

Author: Made this more explicit, and specifically mentioned that resources were heterogeneous.

>   - The response **must not** provide any other information about the issue.

Reviewer: , for example, error details or codes. (To discover what the underlying issue is, users should send a more specific request, e.g., if a single resource is unavailable, users should use a Get request to explore the error in more detail).

Author: Done.

  - The service **must** provide a way for the user to get an error with
    additional information, and **should** allow the user to repeat the
    original call with more restrictive parameters in order to do so.

**Important:** If a single unreachable location or resource prevents returning
any data by definition (for example, a list request for a single publisher
where that publisher's API is down), the service **must** fail the entire
request with an error.

### Pagination

When paginating over a list, it is likely that the service will not know that
there are unreachable locations or resources initially.

- The response **may** provide the unreachable location or resource in the
  `unreachable` field on the page where it is initially encountered.
- The response **must** provide any outstanding unreachable locations or
  resources in the `unreachable` field on the final page.

>   - If a location or resource was unreachable during pagination, but the

Reviewer: I think we still need to be more specific on the pagination situation. For example, you might have parent 1 unavailable, then more pages, then parent 1 comes back. In that scenario, how can we be sure that all the stuff from parent 1 is included? Let's be explicit here about scenarios.

Author: Changed based on the discussion on the call.

---
aip:
  id: 217
  state: reviewing
  created: 2019-08-26
permalink: /217
redirect_from:
  - /0217
---

# Unreachable resources

Occasionally, a user may ask for a list of resources, and some set of resources
in the list are temporarily unavailable. For example, a user may ask to list
resources across multiple locations, but one of those locations is temporarily

> down. In this situation, it is still desirable to give the user as much as

Reviewer: ```suggestion
unavailable. In this situation, it's still desirable to give the user as much as
```

possible, while indicating that something is missing.

## Guidance

> If a method to retrieve data is capable of partially failing due to one or more

Reviewer: I think this might need to be tightened up. It's fine for the reviewing state, but want to be a bit more specific for the final state. I'm concerned about hinting that any retrieval method that can "partially fail" should have a property to show that. For example, what if a Getting a resource fails half-way through when loading data from several sources. It might make more sense to fail the Get request than to provide the partially successful response.

resources being temporarily unreachable, the response message **must** include
a field to indicate this:

```proto
// The request message for ListBooks.

> message ListBooksRequest {

Reviewer: Do we need the request here?

  // The publisher to list books from.
  // To list books across all publishers, use "publishers/-".
  string parent = 1;

// Other fields (as defined in AIP-132)...
}

message ListBooksResponse {
  // The books matching the request.
  repeated Book books = 1;

  // The next page token, if there are more books matching the
  // request.
  string next_page_token = 2;

  // Unreachable resources.
  repeated string unreachable = 3;
}
```

- The field **must** be a repeated string, and **should** be named
  `unreachable`.
- The field **must** be set to the names of the resources which are the cause

>   of the issue, such as the locations or resources that could not be reached.

Reviewer: ```suggestion
of the issue, such as the parent or individual resources that could not be reached.
```

  The objects listed as unreachable **may** be _parents_ (or higher ancestors)
  rather than the individual resources being requested. For example, if a
  location is down, the location is listed.
  - The response **must not** provide any other information about the issue,
    such as error details or codes. To discover what the underlying issue is,
    users **should** send a more specific request.
  - The service **must** provide a way for the user to get an error with
    additional information, and **should** allow the user to repeat the
    original call with more restrictive parameters in order to do so.
  - The resource names provided in this field **may** be heterogeneous. The
    field **should** document what potential resources may be provided in this
    field, and note that it might expand later.

**Important:** If a single unreachable location or resource prevents returning
any data by definition (for example, a list request for a single publisher

> where that publisher's API is down), the service **must** fail the entire

Reviewer: ```suggestion
where that publisher is unavailable), the service **must** fail the entire
```

  `unreachable`.
- The field **must** be set to the names of the resources which are the cause
  of the issue, such as the locations or resources that could not be reached.

>   The objects listed as unreachable **may** be _parents_ rather than the

Reviewer 2: Ancestors? Or do we think "parents" implicitly includes "parents of parents"?

Author: Good point. Done.

  individual resources being requested. For example, if a location is down, the
  location is listed.
  - The response **must not** provide any other information about the issue,
    such as error details or codes. To discover what the underlying issue is,
    users **should** send a more specific request.
  - The service **must** provide a way for the user to get an error with
    additional information, and **should** allow the user to repeat the
    original call with more restrictive parameters in order to do so.

>   - The resource names provided in this field **may** be hetergeneous. The

Reviewer 2: Typo - hetergeneous => heterogeneous

Author: Fixed.

When paginating over a list, it is likely that the service will not know that
there are unreachable locations or resources initially.

- The response **must** provide any outstanding unreachable locations or

>   resources in the `unreachable` field on the _final_ page.

Reviewer 2: It would probably be clearer to write "on pages following the final page that contains resource". There's an implication in "the final page" that there's only one such page.

Author: Done.

  - The response **should not** include both requested data and unreachable
    resources on the same page.
    - For example, if there are two pages of books and one unavailable
      publisher, there should be three pages total: first the two pages of
      books, and then a final page with no books and the unavailable publisher.

>   - If the number of unavailable resources to list is very large, the response

Reviewer 2: With the above suggested change, this would just be "Responses containing unreachable resources should honor the page_size field in the same way as for the resources".

(Note the change from unavailable to unreachable as well.)

Author: Done.
